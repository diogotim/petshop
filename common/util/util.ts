export const arrayFromRange = (start: number, range: number) => {
  return Array.from(Array(range + 1 - start).keys()).map((n) => n + start);
};

export const formatProductSlug = (name: string) => {
  return name.replace(/\s/g, "-").toLowerCase();
};
