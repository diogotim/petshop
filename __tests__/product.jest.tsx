import { render } from "@testing-library/react";
import Product, {
  getStaticPaths,
  getStaticProps,
} from "../pages/product/[slug]";
import mock from "../mock";
import APIRequest from "../modules/api/lib/API";
import { formatProductSlug } from "../common/util/util";
const mockedProduct = mock[0];

const mockedAPI = jest.mocked(APIRequest, true);

describe("Product", () => {
  it("renders product unchanged", () => {
    const prod = render(<Product product={mockedProduct} />);
    expect(prod).toMatchSnapshot();
  });

  it("renders a buy button", () => {
    const { getAllByRole } = render(<Product product={mockedProduct} />);
    expect(getAllByRole("button").length).toEqual(1);
  });
});

describe("Product data fetching", () => {
  beforeEach(() => {
    mockedAPI.mockClear();
  });

  it("getStaticProps returns product data from the API", async () => {
    const jsonRes = () => new Promise((res) => res({ data: mockedProduct }));
    // @ts-ignore
    mockedAPI.mockReturnValue(new Promise((res) => res({ json: jsonRes })));

    const staticProps = await getStaticProps({
      params: { slug: formatProductSlug(mockedProduct.name) },
    });
    expect(staticProps).toStrictEqual({
      props: {
        product: mockedProduct,
      },
    });
  });

  it("getStaticPaths returns product slugs", async () => {
    const jsonRes = () => new Promise((res) => res({ data: [mockedProduct] }));
    // @ts-ignore
    mockedAPI.mockReturnValue(new Promise((res) => res({ json: jsonRes })));

    const staticPaths = await getStaticPaths();
    expect(staticPaths).toStrictEqual({
      paths: [{ params: { slug: formatProductSlug(mockedProduct.name) } }],
      fallback: false,
    });
  });
});
