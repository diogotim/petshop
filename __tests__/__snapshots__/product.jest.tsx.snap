// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product renders product unchanged 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header>
        <div
          class="header container"
        >
          <a
            class="logo"
            href="/"
          >
            Pets
          </a>
          <nav
            class="nav"
          >
            <ul>
              <li>
                <a
                  href="/"
                >
                  Adoption
                </a>
              </li>
              <li>
                <a
                  href="/"
                >
                  Services
                </a>
              </li>
              <li>
                <a
                  href="/products/1"
                >
                  Products
                </a>
              </li>
            </ul>
          </nav>
        </div>
      </header>
      <main>
        <div
          class="container root"
        >
          <h1>
            Beautiful Scarf
          </h1>
          <div
            class="productDetails"
          >
            <div
              class="pic"
            >
              <img
                alt="product image"
                class="imgContain imgFill"
                src="/beautiful-scarf.jpg"
              />
            </div>
            <div
              class="info"
            >
              <p
                class="price"
              >
                Available for $
                213.44
              </p>
              <button
                class="buyBtn"
                type="button"
              >
                Buy Now
              </button>
            </div>
          </div>
        </div>
      </main>
      <footer
        class="root"
      >
        <div
          class="footer container"
        >
          <a
            class="logo"
            href="/"
          >
            Pets
          </a>
          <nav
            class="nav"
          >
            <ul>
              <li>
                <p>
                  Quick links
                </p>
                <a
                  href="/"
                >
                  Adoption
                </a>
                <a
                  href="/"
                >
                  Services
                </a>
                <a
                  href="/products/1"
                >
                  Products
                </a>
              </li>
              <li>
                <p>
                  Follow us
                </p>
                <a
                  href="#"
                >
                  Facebook
                </a>
                <a
                  href="#"
                >
                  Instagram
                </a>
                <a
                  href="#"
                >
                  Twitter
                </a>
              </li>
              <li>
                <p>
                  Contact
                </p>
                <a
                  href="#"
                >
                  +0 (000) 000-0000
                </a>
                <a
                  href="#"
                >
                  someone@example.com
                </a>
              </li>
            </ul>
          </nav>
        </div>
      </footer>
    </div>
  </body>,
  "container": <div>
    <header>
      <div
        class="header container"
      >
        <a
          class="logo"
          href="/"
        >
          Pets
        </a>
        <nav
          class="nav"
        >
          <ul>
            <li>
              <a
                href="/"
              >
                Adoption
              </a>
            </li>
            <li>
              <a
                href="/"
              >
                Services
              </a>
            </li>
            <li>
              <a
                href="/products/1"
              >
                Products
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </header>
    <main>
      <div
        class="container root"
      >
        <h1>
          Beautiful Scarf
        </h1>
        <div
          class="productDetails"
        >
          <div
            class="pic"
          >
            <img
              alt="product image"
              class="imgContain imgFill"
              src="/beautiful-scarf.jpg"
            />
          </div>
          <div
            class="info"
          >
            <p
              class="price"
            >
              Available for $
              213.44
            </p>
            <button
              class="buyBtn"
              type="button"
            >
              Buy Now
            </button>
          </div>
        </div>
      </div>
    </main>
    <footer
      class="root"
    >
      <div
        class="footer container"
      >
        <a
          class="logo"
          href="/"
        >
          Pets
        </a>
        <nav
          class="nav"
        >
          <ul>
            <li>
              <p>
                Quick links
              </p>
              <a
                href="/"
              >
                Adoption
              </a>
              <a
                href="/"
              >
                Services
              </a>
              <a
                href="/products/1"
              >
                Products
              </a>
            </li>
            <li>
              <p>
                Follow us
              </p>
              <a
                href="#"
              >
                Facebook
              </a>
              <a
                href="#"
              >
                Instagram
              </a>
              <a
                href="#"
              >
                Twitter
              </a>
            </li>
            <li>
              <p>
                Contact
              </p>
              <a
                href="#"
              >
                +0 (000) 000-0000
              </a>
              <a
                href="#"
              >
                someone@example.com
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
